### ISSUE TITLE
🐌  Slow tests

### ISSUE THREAD

**Issue by blaginin** [TissueImageAnalytics/tiatoolbox#603]
- TIA Toolbox version: 1.4.0
- Python version:  3.10.8 
- Operating System: Mac OS 13.3.1 (arm)

### Description

Tests in the toolbox are extremely slow. Even with `CI=True`, all the checks take almost half an hour. Non-CI mode takes even more time. This makes an interactive development approach, with running tests after small changes, impossible. This results in writing [bad](https://github.com/blaginin/tiatoolbox/commit/267b91af795ccc548c8a19a9026004e97510bbfa#diff-5c6dddc0c85981eb75bdb49d400e82c9271918e138990baba53ac34cbcf2aaacR29) test-related code.

The main reasons why tests are slow are static files. Each test run is related to downloading and then deleting gigabytes of models' weights and test images. However, almost always, those files remain unchanged after the tests' execution. 

Not only is this slow, but it also triggers a lot of false positive results. For example, that's the typical result of my local run: 

<img width="1439" alt="image" src="https://user-images.githubusercontent.com/19199204/236876594-908c7835-704f-4ad6-938a-f761796b49dd.png">

# Todo

- [x] Fixtures caching
- [x] Models caching
- [x] Cache in CI
- [x] Parallel tests
- [ ] Safe downloads
- [ ] Cache rechecks (HTTP headers / ETag)

**Comment by blaginin** [comment]
Of course, we want to be sure of the tests' reproducibility and don't want our code to fail because some code before has modified our assets. But maybe we can consider more user-friendly approaches? I have those options in mind: 
- Compare a file's local last change time with the remote one and download only if it's greater. This is done by `If-Range` / `Last-Modified` headers, which are already supported by `tiatoolbox.dcs.warwick.ac.uk` server. 
- Store files checksum on the server side and download a file only if its checksum is different
- Copy files from the local cache folder into the tests' sandbox each time tests are run

**Comment by blaginin** [comment]
@John-P @shaneahmed, maybe you have other ideas?

**Comment by shaneahmed** [comment]
@blaginin As John mentioned downloads should be for a session. Please let us know if this is not the case. Feel free to create a PR to fix this.
